specs/vep-1416-vdk-troubleshooting-tools/README.md

```diff
@@ -0,0 +1,101 @@
+
+In such cases, there are a couple of approaches that can be taken:
+
+1. An execution of the job can be triggered manually, and the administrator can connect to the cluster directly and either examine the logs as they are produced, or attach to the pod of the data job and troubleshoot it remotely. This has the limitation that the person doing the troubleshooting needs to have access to the kubernetes cluster, and even then the container in which the data job is running may not have all the utilities (editor, program for monitoring of running processes, etc.) to allow for proper troubleshooting. Additionally, it may not be possible for the data job to be run off its schedule due to business concerns, in which case the users/admins would need to wait for the job's next scheduled execution.
+2. The data job can be executed locally, so that the user/admin can have complete control over the environment. This again has the drawback that off-schedule execution of the data job may not be possible. Also, in case the issue is with some environment configuration or system dependency, the error may not be reproducible locally, which can make troubleshooting even more difficult.
+
+To allow for easier troubleshooting of errors with data jobs, a special plugin will be introduced, which will provide capabilities to do a thread dump and either log the contents or stored locally, where it can be examined by the owner of the data job or by the team responsible for the VDK deployment. The plugin could also be extended in the future with other debugging capabilities.
+
+
+## Requirements and goals
+
+-----
+### Goals
+
+* **Introduce a vdk-jobs-troubleshooting plugin**
+
+* **Provide capability to do a thread dump**
+  - An administrator wants to get a thread dump of the data job process for troubleshooting purposes. For example, a user deploys a data job and after some time the job starts failing without any apparent reason. The administrator needs to be able to get a thread dump, so they are able to properly investigate the root cause for the issue.
+
+### Non-Goals
+
+* **Additional troubleshooting tools are not planned as part of this proposal.**
+* **The thread-dump feature would not be exposed to users/data job owners.**
+
+
+## High-level design
+
+-----
+
+![plugin_diagram.png](diagrams/plugin_diagram.png)
+
+For the proposed design, a vdk-jobs-troubleshooting plugin will be introduced. The plugin will start a http server instance, which will allow users to obtain a thread dump (which will also be printed in the logs) for further troubleshooting.
+
+
+## API design
+
+-----
+
+No changes to the public API.
+
```

Comment:
Is there any change on the k8s pod interface surface? Where is the HTTP server listening / awaiting requests? What is a valid request?

Reply:
The HTTP server will be listening on localhost, so no port or anything will be exposed to the outside world. The connection will happen when an administrator with access to the k8s cluster does a port-forward to the data job pod.
Will provide more details in a subsequent PR.
Thank you for the review